server/src/main/java/org/elasticsearch/ingest/IngestStats.java

```diff
@@ -242,19 +245,56 @@ IngestStats build() {
     /**
      * Container for pipeline stats.
      */
-    public record PipelineStat(String pipelineId, Stats stats) {
+    public record PipelineStat(String pipelineId, Stats stats, ByteStats byteStats) {
         static List<PipelineStat> merge(List<PipelineStat> first, List<PipelineStat> second) {
-            var totalsPerPipeline = new HashMap<String, Stats>();
+            var totalsPerPipeline = new HashMap<String, PipelineStat>();
 
-            first.forEach(ps -> totalsPerPipeline.merge(ps.pipelineId, ps.stats, Stats::merge));
-            second.forEach(ps -> totalsPerPipeline.merge(ps.pipelineId, ps.stats, Stats::merge));
+            first.forEach(ps -> totalsPerPipeline.merge(ps.pipelineId, ps, PipelineStat::merge));
+            second.forEach(ps -> totalsPerPipeline.merge(ps.pipelineId, ps, PipelineStat::merge));
 
             return totalsPerPipeline.entrySet()
                 .stream()
-                .map(v -> new PipelineStat(v.getKey(), v.getValue()))
+                .map(v -> new PipelineStat(v.getKey(), v.getValue().stats, v.getValue().byteStats))
                 .sorted(PIPELINE_STAT_COMPARATOR)
                 .toList();
         }
+
+        private static PipelineStat merge(PipelineStat first, PipelineStat second) {
+            assert first.pipelineId.equals(second.pipelineId) : "Can only merge stats from the same pipeline";
+            return new PipelineStat(
+                first.pipelineId,
+                Stats.merge(first.stats, second.stats),
+                ByteStats.merge(first.byteStats, second.byteStats)
+            );
+        }
+    }
+
+    /**
+     * Container for ingested byte stats
+     */
+    public record ByteStats(long bytesIngested, long bytesProduced) implements Writeable, ToXContentFragment {
+        public ByteStats(StreamInput in) throws IOException {
+            this(in.readVLong(), in.readVLong());
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeVLong(bytesIngested);
+            out.writeVLong(bytesProduced);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            if (bytesIngested > 0 || bytesProduced > 0) {
```

Comment:
Having optional fields in the response is kind of a pain for clients to handle. Possible ofc, just more hassle than having them always present. The clients team have expressed a desire to avoid this wherever possible. Could we follow up with a change to emit these fields always?

Reply:
@DaveCTurner That makes sense that it is a pain for clients to handle. We were worried that because there a several situations where a pipeline would only have 0 values (the final pipeline after a default pipeline, a pipeline processor, after a reroute) and these bytes are added to the associated first pipeline, it would be confusing to have a 0 value. Leaving the values out of the response seemed a way to reduce confusion.

Could using optional response fields might be warranted in this situation to reduce confusion?

Reply:
If we were genuinely distinguishing between these cases then mmmmaybe. But we're not, we're just suppressing these values if they're zero.

A pipeline might be the first pipeline for some docs and a chained pipeline for other docs, and then we're going to report these numbers but only for the docs for which the pipeline wasn't chained from some earlier one. So the confusion exists anyway IMO.

Reply:
Yeah, it really is confusing either way :sweat_smile:  Sounds good, I'll make that change.

Reply:
Great, thanks

Reply:
(we could perhaps come up with less generic names to avoid folks thinking that `bytes_ingested` is the total size of all documents processed by this pipeline, an obvious interpretation otherwise...)

Reply:
> `ingested_as_first_pipeline_in_bytes` ? 

Yep something like that would be fine with me.

> maybe it's just confusing enough so that users decide they should really read the docs

Exactly.

Also the `_in_bytes` thing is another good point, we should add human-readable versions of these values too:

```java
builder.humanReadableField("ingested_as_first_pipeline_in_bytes", "ingested_as_first_pipeline", ByteSizeValue.ofBytes(bytesIngested));
```

Reply:
+1 to adding human readable versions for these.